### ISSUE TITLE
figure out why unused functions from builtin.a are not getting omitted from final binary

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2062]
I'm pretty sure when I solved #54 it was working but now if I do this:

```
$ ./zig build-exe ../examples/hello_world/hello.zig --release-small
$ objdump -d hello
...
0000000000201a40 <sqrt>:
...
0000000000201bf0 <sqrtf>:
...
```

That's not supposed to be in there. `sqrt` (among others) isn't called from this application.

On the other hand functions from compiler_rt.a are correctly getting omitted.

Maybe we should do `-ffunction-sections`?

**Comment by vegecode** [comment]
It seems to happen only in release-safe and release-fast modes for me when compiling for
armv7em-freestanding-eabihf.

In debug and release small modes, there are no calls to any compiler-rt functions that I can see.

**Comment by andrewrk** [comment]
Note for webassembly what @shritesh found: https://github.com/ziglang/zig/pull/2248#issuecomment-481904959

**Comment by vegecode** [comment]
I just pulled the latest zig and my freestanding arm project won't build anymore:

```
lld: error: undefined symbol: __zig_probe_stack
>>> referenced by debug.zig:1822 (/home/alexander/Code/zig/build/lib/zig/std/debug.zig:1822)
>>>               /home/alexander/Code/msp432-tiva-zig/zig-cache/o/siMKlSLAjLobrJZ7TlKJUPyWP9vPyHsp5b0NmB2ajmfniXlyxalD_J1uFQo1EQ7b/build.o:(std.debug.getLineNumberInfoDwarf)
```

I think it is related to this issue.

**Comment by timonkruiper** [comment]
I'm pretty sure that the problem is that all the different compiler_rt functions all end up in one object file (compiler_rt.o). In that case the linker has no option to link the complete object and thus end up with all the unnecessary code in the final executable. 

Instead to solve this problem all the different zig files in the compiler_rt should end up in a separate object file inside the libcompiler_rt.a This way the linker can choose the correct object file.

I found this by doing `ar -t` on the libcompiler_rt.a in the .local/share/zig cache folder and by doing the same command on the libclang_rt.a file.

**Comment by timonkruiper** [comment]
@andrewrk I would like to solve this, because it's pretty annoying when build stuff for small embedded microcontrollers. However, I'm not sure if I should implement `-ffunction-sections` or go with the method described above?

**Comment by andrewrk** [comment]
I would try `-ffunction-sections` first since in theory it should just be an option provided to LLVM. Getting compiler_rt to be more than 1 object file would be a lot more work and probably make compilation slower.

**Comment by andrewrk** [comment]
Landed in 372b615ace0d92d51521b8b0e347d90400f8f7a1. Thanks @FireFox317! Hello World on Linux is now 6 KB after being stripped of debug info:

```
[nix-shell:~/downloads/zig/build]$ ./zig build-exe ../example/hello_world/hello.zig --release-small --strip --single-threaded

[nix-shell:~/downloads/zig/build]$ ls -ahl hello
-rwxr-xr-x 1 andy users 9.8K Jul  3 15:49 hello

[nix-shell:~/downloads/zig/build]$ strip hello # zig's --strip isn't fully complete yet

[nix-shell:~/downloads/zig/build]$ ls -ahl hello
-rwxr-xr-x 1 andy users 6.0K Jul  3 15:49 hello

[nix-shell:~/downloads/zig/build]$ ./hello
Hello, world!
```
